File: ERP Claude K/frontend/lib/core/services/socket_service.dart
import 'package:flutter/foundation.dart';
import 'package:socket_io_client/socket_io_client.dart' as socket_io;
import 'package:enterprise_erp/core/constants/app_constants.dart';
import 'package:enterprise_erp/core/utils/storage_service.dart';

class SocketService {
  static final SocketService _instance = SocketService._internal();
  factory SocketService() => _instance;
  SocketService._internal();

  socket_io.Socket? _socket;
  final StorageService _storage = StorageService();
  bool _isConnected = false;

  bool get isConnected => _isConnected;

  // Callbacks
  final Map<String, List<Function>> _eventListeners = {};

  Future<void> connect() async {
    if (_socket != null && _socket!.connected) {
      debugPrint('Socket already connected');
      return;
    }

    try {
      final token = await _storage.getAuthToken();
      if (token == null) {
        debugPrint('No auth token found for socket connection');
        return;
      }

      _socket = socket_io.io(
        AppConfig.socketUrl,
        socket_io.OptionBuilder()
            .setTransports(['websocket'])
            .disableAutoConnect()
            .setAuth({'token': token})
            .setExtraHeaders({'Authorization': 'Bearer $token'})
            .build(),
      );

      _socket!.onConnect((_) {
        debugPrint('✅ Socket connected');
        _isConnected = true;
        _notifyListeners('connected', null);
      });

      _socket!.onDisconnect((_) {
        debugPrint('❌ Socket disconnected');
        _isConnected = false;
        _notifyListeners('disconnected', null);
      });

      _socket!.onError((error) {
        debugPrint('Socket error: $error');
        _notifyListeners('error', error);
      });

      // Chat events
      _socket!.on('new_message', (data) {
        debugPrint('New message received: $data');
        _notifyListeners('new_message', data);
      });

      _socket!.on('user_typing', (data) {
        _notifyListeners('user_typing', data);
      });

      _socket!.on('message_read', (data) {
        _notifyListeners('message_read', data);
      });

      _socket!.on('user_status_changed', (data) {
        _notifyListeners('user_status_changed', data);
      });

      _socket!.on('rooms_joined', (data) {
        debugPrint('Rooms joined: $data');
        _notifyListeners('rooms_joined', data);
      });

      _socket!.on('new_expense', (data) {
        debugPrint('New expense: $data');
        _notifyListeners('new_expense', data);
      });

      // Video call events
      _socket!.on('incoming_call', (data) {
        _notifyListeners('incoming_call', data);
      });

      _socket!.on('call_answered', (data) {
        _notifyListeners('call_answered', data);
      });

      _socket!.on('call_ended', (data) {
        _notifyListeners('call_ended', data);
      });

      _socket!.on('ice_candidate', (data) {
        _notifyListeners('ice_candidate', data);
      });

      _socket!.connect();
    } catch (e) {
      debugPrint('Socket connection error: $e');
    }
  }

  void disconnect() {
    if (_socket != null) {
      _socket!.disconnect();
      _socket!.dispose();
      _socket = null;
      _isConnected = false;
      debugPrint('Socket disconnected manually');
    }
  }

  // Join chat rooms
  void joinRooms(List<String> roomIds) {
    if (_socket != null && _socket!.connected) {
      _socket!.emit('join_rooms', roomIds);
      debugPrint('Joining rooms: $roomIds');
    }
  }

  // Leave chat room
  void leaveRoom(String roomId) {
    if (_socket != null && _socket!.connected) {
      _socket!.emit('leave_room', roomId);
      debugPrint('Left room: $roomId');
    }
  }

  // Send typing indicator
  void sendTyping(String roomId, bool isTyping) {
    if (_socket != null && _socket!.connected) {
      _socket!.emit('typing', {'roomId': roomId, 'isTyping': isTyping});
    }
  }

  // Mark message as read
  void markAsRead(String messageId, String roomId) {
    if (_socket != null && _socket!.connected) {
      _socket!.emit('mark_read', {'messageId': messageId, 'roomId': roomId});
    }
  }

  // Update online status
  void updateStatus(String status) {
    if (_socket != null && _socket!.connected) {
      _socket!.emit('update_status', status);
    }
  }

  // Video call signaling
  void callUser(String targetUserId, dynamic offer) {
    if (_socket != null && _socket!.connected) {
      _socket!.emit('call_user', {'targetUserId': targetUserId, 'offer': offer});
    }
  }

  void answerCall(String targetUserId, dynamic answer) {
    if (_socket != null && _socket!.connected) {
      _socket!.emit('answer_call', {'targetUserId': targetUserId, 'answer': answer});
    }
  }

  void sendIceCandidate(String targetUserId, dynamic candidate) {
    if (_socket != null && _socket!.connected) {
      _socket!.emit('ice_candidate', {'targetUserId': targetUserId, 'candidate': candidate});
    }
  }

  void endCall(String targetUserId) {
    if (_socket != null && _socket!.connected) {
      _socket!.emit('end_call', {'targetUserId': targetUserId});
    }
  }

  // Event listener management
  void addEventListener(String event, Function callback) {
    if (!_eventListeners.containsKey(event)) {
      _eventListeners[event] = [];
    }
    _eventListeners[event]!.add(callback);
  }

  void removeEventListener(String event, Function callback) {
    if (_eventListeners.containsKey(event)) {
      _eventListeners[event]!.remove(callback);
    }
  }

  void removeAllListeners(String event) {
    _eventListeners.remove(event);
  }

  void _notifyListeners(String event, dynamic data) {
    if (_eventListeners.containsKey(event)) {
      for (var callback in _eventListeners[event]!) {
        callback(data);
      }
    }
  }
}

File: ERP Claude K/backend/services/socketService.js
const socketIO = require('socket.io');
const jwt = require('jsonwebtoken');
const User = require('../models/User');
const { ChatRoomMember } = require('../models/Chat');

const initializeSocket = (server) => {
  const io = socketIO(server, {
    cors: {
      origin: process.env.CORS_ORIGIN.split(','),
      methods: ['GET', 'POST'],
      credentials: true
    },
    pingTimeout: 60000,
    pingInterval: 25000
  });

  // Authentication middleware
  io.use(async (socket, next) => {
    try {
      const token = socket.handshake.auth.token || socket.handshake.headers.authorization?.split(' ')[1];
      
      if (!token) {
        return next(new Error('Authentication token required'));
      }

      const decoded = jwt.verify(token, process.env.JWT_SECRET);
      
      const user = await User.findByPk(decoded.id);
      if (!user || !user.is_active) {
        return next(new Error('Invalid or inactive user'));
      }

      socket.userId = user.id;
      socket.companyId = user.company_id;
      socket.username = user.username;
      
      next();
    } catch (error) {
      next(new Error('Authentication failed'));
    }
  });

  // Connection handler
  io.on('connection', (socket) => {
    console.log(`✅ User connected: ${socket.username} (${socket.userId})`);

    // Join user's personal room
    socket.join(`user:${socket.userId}`);
    socket.join(`company:${socket.companyId}`);

    // Join chat rooms
    socket.on('join_rooms', async (roomIds) => {
      try {
        // Verify user is member of these rooms
        const memberships = await ChatRoomMember.findAll({
          where: {
            user_id: socket.userId,
            room_id: roomIds
          }
        });

        const validRoomIds = memberships.map(m => m.room_id);
        
        validRoomIds.forEach(roomId => {
          socket.join(roomId);
          console.log(`User ${socket.username} joined room ${roomId}`);
        });

        socket.emit('rooms_joined', { rooms: validRoomIds });
      } catch (error) {
        console.error('Join rooms error:', error);
        socket.emit('error', { message: 'Failed to join rooms' });
      }
    });

    // Leave chat room
    socket.on('leave_room', (roomId) => {
      socket.leave(roomId);
      console.log(`User ${socket.username} left room ${roomId}`);
    });

    // Typing indicator
    socket.on('typing', ({ roomId, isTyping }) => {
      socket.to(roomId).emit('user_typing', {
        userId: socket.userId,
        username: socket.username,
        roomId,
        isTyping
      });
    });

    // Mark message as read
    socket.on('mark_read', ({ messageId, roomId }) => {
      socket.to(roomId).emit('message_read', {
        messageId,
        userId: socket.userId,
        timestamp: new Date()
      });
    });

    // Online status
    socket.on('update_status', (status) => {
      io.to(`company:${socket.companyId}`).emit('user_status_changed', {
        userId: socket.userId,
        status,
        timestamp: new Date()
      });
    });

    // Video call signaling
    socket.on('call_user', ({ targetUserId, offer }) => {
      io.to(`user:${targetUserId}`).emit('incoming_call', {
        from: socket.userId,
        username: socket.username,
        offer
      });
    });

    socket.on('answer_call', ({ targetUserId, answer }) => {
      io.to(`user:${targetUserId}`).emit('call_answered', {
        from: socket.userId,
        answer
      });
    });

    socket.on('ice_candidate', ({ targetUserId, candidate }) => {
      io.to(`user:${targetUserId}`).emit('ice_candidate', {
        from: socket.userId,
        candidate
      });
    });

    socket.on('end_call', ({ targetUserId }) => {
      io.to(`user:${targetUserId}`).emit('call_ended', {
        from: socket.userId
      });
    });

    // Disconnect handler
    socket.on('disconnect', () => {
      console.log(`❌ User disconnected: ${socket.username}`);
      
      // Notify others in the company
      io.to(`company:${socket.companyId}`).emit('user_status_changed', {
        userId: socket.userId,
        status: 'offline',
        timestamp: new Date()
      });
    });

    // Error handler
    socket.on('error', (error) => {
      console.error(`Socket error for user ${socket.username}:`, error);
    });
  });

  // Store io instance globally for use in controllers
  global.io = io;

  return io;
};

module.exports = { initializeSocket };
